File: backend/python-api/app/service/analytics_service.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone

from app.repository.analytics_repo import AnalyticsRepository


class AnalyticsService:
    def __init__(self, analytics_repo: AnalyticsRepository):
        self._analytics = analytics_repo

    async def user_registrations(
        self,
        start: datetime | None = None,
        end: datetime | None = None,
        country: str | None = None,
        language: str | None = None,
    ) -> list[dict]:
        if end is None:
            end = datetime.now(timezone.utc)
        if start is None:
            start = end - timedelta(days=30)
        return await self._analytics.user_registrations_by_day(start, end, country, language)

    async def active_users(self, start: datetime | None = None, end: datetime | None = None) -> dict:
        now = datetime.now(timezone.utc)
        if end is None:
            end = now
        if start is None:
            start = end - timedelta(days=30)

        daily = await self._analytics.active_users_by_day(start, end)
        active_24h = await self._analytics.active_users_count(now - timedelta(hours=24))
        active_7d = await self._analytics.active_users_count(now - timedelta(days=7))

        return {"daily": daily, "active_24h": active_24h, "active_7d": active_7d}

    async def retention(self, start: datetime | None = None, end: datetime | None = None) -> dict:
        if end is None:
            end = datetime.now(timezone.utc)
        if start is None:
            start = end - timedelta(days=90)

        data = await self._analytics.retention_by_day(start, end)
        avg_7d = sum(d["retention_7_rate"] for d in data) / len(data) if data else 0.0
        avg_30d = sum(d["retention_30_rate"] for d in data) / len(data) if data else 0.0

        return {"retention_data": data, "average_7d": round(avg_7d, 2), "average_30d": round(avg_30d, 2)}

    async def cohort_retention(self, start: datetime | None = None, end: datetime | None = None) -> list[dict]:
        if end is None:
            end = datetime.now(timezone.utc)
        if start is None:
            start = end - timedelta(days=90)
        return await self._analytics.cohort_retention(start, end)

    async def summary(self) -> dict:
        return await self._analytics.summary_stats()

File: backend/go-api/internal/service/analytics_service.go
package service

import (
	"context"
	"time"

	"saas-starter/backend/go-api/internal/dto"
	"saas-starter/backend/go-api/internal/repository"
)

type AnalyticsService struct {
	analytics repository.AnalyticsRepository
}

func NewAnalyticsService(analytics repository.AnalyticsRepository) *AnalyticsService {
	return &AnalyticsService{analytics: analytics}
}

func (s *AnalyticsService) UserRegistrations(ctx context.Context, filter dto.AnalyticsFilter) ([]dto.UserAnalytics, error) {
	return s.analytics.UserRegistrationsByDay(ctx, filter)
}

func (s *AnalyticsService) ActiveUsers(ctx context.Context, filter dto.AnalyticsFilter) (*dto.ActiveUserAnalyticsResponse, error) {
	daily, err := s.analytics.ActiveUsersByDay(ctx, filter)
	if err != nil {
		return nil, err
	}

	active24h, _ := s.analytics.ActiveUsersCount(ctx, time.Now().Add(-24*time.Hour), filter)
	active7d, _ := s.analytics.ActiveUsersCount(ctx, time.Now().AddDate(0, 0, -7), filter)

	return &dto.ActiveUserAnalyticsResponse{
		Daily:     daily,
		Active24h: active24h,
		Active7d:  active7d,
	}, nil
}

func (s *AnalyticsService) Retention(ctx context.Context, filter dto.AnalyticsFilter) (*dto.RetentionAnalyticsResponse, error) {
	data, err := s.analytics.RetentionByDay(ctx, filter)
	if err != nil {
		return nil, err
	}

	var avg7, avg30 float64
	count := 0
	for _, r := range data {
		if r.NewUsers > 0 {
			avg7 += r.Retention7Rate
			avg30 += r.Retention30Rate
			count++
		}
	}
	if count > 0 {
		avg7 /= float64(count)
		avg30 /= float64(count)
	}

	return &dto.RetentionAnalyticsResponse{
		RetentionData: data,
		Average7d:     avg7,
		Average30d:    avg30,
	}, nil
}

func (s *AnalyticsService) CohortRetention(ctx context.Context, filter dto.AnalyticsFilter) ([]dto.CohortRow, error) {
	return s.analytics.CohortRetention(ctx, filter)
}

func (s *AnalyticsService) Summary(ctx context.Context, filter dto.AnalyticsFilter) (*dto.AdminSummary, error) {
	return s.analytics.SummaryStats(ctx, filter)
}
